### PR TITLE
TPCClusterFinder: Remove duplicate cluster native and fix peak finding

### PR DESCRIPTION
**Pull request by fweig**
This Pull Request...
* removes the duplicate ClusterNative class that was leftover from the standalone cluster finder
* replaces several macros with templates
* fixes a bug where the cluster finder would find adjacent peaks
* and a bit more refactoring

@davidrohr

### CODE REVIEW

File: GPU/GPUTracking/TPCClusterFinder/ClusterAccumulator.cxx [hunk 1]
```diff
@@ -13,67 +13,62 @@
 
 #include "ClusterAccumulator.h"
 #include "CfUtils.h"
-#include "ClusterNative.h"
```
Comment (davidrohr) [inline GPU/GPUTracking/TPCClusterFinder/ClusterAccumulator.cxx]: This change (and some other) triggered some warnings during compilation. See below. Could you fix these:

```
/home/qon/alice/O2/GPU/GPUTracking/Standalone/TPCClusterFinder/GPUTPCClusterFinder.h:26:1: warning: class 'ClusterNative' was previously declared as a struct; this is valid, but may result in linker errors under the Microsoft C++ ABI [-Wmismatched-tags]
class ClusterNative;
^
/home/qon/alice/O2/GPU/GPUTracking/Standalone/../../../DataFormats/Detectors/TPC/include/DataFormatsTPC/ClusterNative.h:50:8: note: previous use is here
struct ClusterNative {
       ^
/home/qon/alice/O2/GPU/GPUTracking/Standalone/TPCClusterFinder/GPUTPCClusterFinder.h:26:1: note: did you mean struct here?
class ClusterNative;
^~~~~
struct                                                                                                                                                                                                                                                                                                                                                     
In file included from /home/qon/alice/O2/GPU/GPUTracking/Base/hip/GPUReconstructionHIP.hip.cxx:62:
In file included from /home/qon/alice/O2/GPU/GPUTracking/Standalone/Base/GPUReconstructionIncludesDevice.h:69:
In file included from /home/qon/alice/O2/GPU/GPUTracking/Standalone/TPCClusterFinder/GPUTPCClusterFinderKernels.cxx:19:
/home/qon/alice/O2/GPU/GPUTracking/Standalone/TPCClusterFinder/Array2D.h:79:1: warning: 'GridSize' defined as a struct template here but previously declared as a class template; this is valid, but may result in linker errors under the Microsoft C++ ABI [-Wmismatched-tags]
struct GridSize<1> {
^
/home/qon/alice/O2/GPU/GPUTracking/Standalone/TPCClusterFinder/Array2D.h:76:1: note: did you mean struct here?
class GridSize;
^~~~~
struct                                                                                                                                                                                                                                                                                                                                                     
/home/qon/alice/O2/GPU/GPUTracking/Standalone/TPCClusterFinder/Array2D.h:87:1: warning: 'GridSize' defined as a struct template here but previously declared as a class template; this is valid, but may result in linker errors under the Microsoft C++ ABI [-Wmismatched-tags]
struct GridSize<2> {
^
/home/qon/alice/O2/GPU/GPUTracking/Standalone/TPCClusterFinder/Array2D.h:76:1: note: did you mean struct here?
class GridSize;
^~~~~
struct                                                                                                                                                                                                                                                                                                                                                     
In file included from /home/qon/alice/O2/GPU/GPUTracking/Base/hip/GPUReconstructionHIP.hip.cxx:62:
In file included from /home/qon/alice/O2/GPU/GPUTracking/Standalone/Base/GPUReconstructionIncludesDevice.h:69:
In file included from /home/qon/alice/O2/GPU/GPUTracking/Standalone/TPCClusterFinder/GPUTPCClusterFinderKernels.cxx:25:
/home/qon/alice/O2/GPU/GPUTracking/Standalone/TPCClusterFinder/ClusterAccumulator.h:25:1: warning: class 'ClusterNative' was previously declared as a struct; this is valid, but may result in linker errors under the Microsoft C++ ABI [-Wmismatched-tags]
class ClusterNative;
^
/home/qon/alice/O2/GPU/GPUTracking/Standalone/../../../DataFormats/Detectors/TPC/include/DataFormatsTPC/ClusterNative.h:50:8: note: previous use is here
struct ClusterNative {
       ^
/home/qon/alice/O2/GPU/GPUTracking/Standalone/TPCClusterFinder/ClusterAccumulator.h:25:1: note: did you mean struct here?
class ClusterNative;
^~~~~
struct                                                                                                                                                                                                                                                                                                                                                     
In file included from /home/qon/alice/O2/GPU/GPUTracking/Base/hip/GPUReconstructionHIP.hip.cxx:62:
In file included from /home/qon/alice/O2/GPU/GPUTracking/Standalone/Base/GPUReconstructionIncludesDevice.h:69:
In file included from /home/qon/alice/O2/GPU/GPUTracking/Standalone/TPCClusterFinder/GPUTPCClusterFinderKernels.cxx:26:
/home/qon/alice/O2/GPU/GPUTracking/Standalone/TPCClusterFinder/Clusterizer.h:27:1: warning: class 'ClusterNative' was previously declared as a class; this is valid, but may result in linker errors under the Microsoft C++ ABI [-Wmismatched-tags]
class ClusterNative;
^
/home/qon/alice/O2/GPU/GPUTracking/Standalone/../../../DataFormats/Detectors/TPC/include/DataFormatsTPC/ClusterNative.h:50:8: note: previous use is here
struct ClusterNative {
       ^
/home/qon/alice/O2/GPU/GPUTracking/Standalone/TPCClusterFinder/Clusterizer.h:27:1: note: did you mean struct here?
class ClusterNative;
^~~~~
struct                                                                            
```

File: GPU/GPUTracking/TPCClusterFinder/PeakFinder.cxx [hunk 2]
```diff
@@ -68,37 +62,40 @@ GPUd() bool PeakFinder::isPeakScratchPad(
...
-  const GlobalPad gpad = Array2D::tpcGlobalPadIdx(row, pad);
-
   bool peak = true;
 
-#define CMP_NEIGHBOR(dp, dt, cmpOp)                           \
-  do {                                                        \
-    PackedCharge p = CHARGE(chargeMap, gpad + dp, time + dt); \
-    const Charge otherCharge = p.unpack();                    \
-    peak &= (otherCharge cmpOp myCharge);                     \
+#define CMP_NEIGHBOR(dp, dt, cmpOp)                  \
```
Comment (davidrohr) [inline GPU/GPUTracking/TPCClusterFinder/PeakFinder.cxx]: This change triggers for some reason an internal compiler error in clang 10 C++ for OpenCL.
I am filing a bug report to clang, but I'll wait with merging until I get a response.

File: GPU/GPUTracking/TPCClusterFinder/ChargeMapFiller.cxx [hunk 3]
```diff
@@ -12,14 +12,15 @@
...
 #include "ChargeMapFiller.h"
+#include "ChargePos.h"
 #include "Array2D.h"
 
 using namespace GPUCA_NAMESPACE::gpu;
 using namespace GPUCA_NAMESPACE::gpu::deprecated;
 
 GPUd() void ChargeMapFiller::fillChargeMapImpl(int nBlocks, int nThreads, int iBlock, int iThread, GPUTPCClusterFinderKernels::GPUTPCSharedMemory& smem,
                                                GPUglobalref() const Digit* digits,
-                                               GPUglobalref() PackedCharge* chargeMap,
```
Comment (davidrohr) [inline GPU/GPUTracking/TPCClusterFinder/ChargeMapFiller.cxx]: In principal, you can remove all instances of GPUglobalref() and also GPUsharedref() in here.
They were just needed for the old OpenCL C code.

Comment (davidrohr) [reply GPU/GPUTracking/TPCClusterFinder/ChargeMapFiller.cxx]: Just to make this more clear: you may only remove GPUglobalref() and GPUsharedref(), the other GPU...ref() must remain. Also, you can only remove them inside the TPCClusterFinder. Part of the other code is still compiler with OpenCL C and needs these macros.

**Comment by davidrohr** [comment]
@fweig : Please apply the attached patch as well, which fixes one of the problem with OpenCL addresses: 
[opencl.patch.gz](https://github.com/AliceO2Group/AliceO2/files/4156115/opencl.patch.gz)
(The root cause is complicated, but the problem is that OpenCL distinguishes between plain ptrs and generic ptrs (even though they are the same), but semantic-wise they are net, and without the GPUgeneric() in there, the template argument deduction fails.

**Comment by davidrohr** [comment]
@fweig : There are more problems in OpenCL with the function pointers. You have 2 of them:
`static GPUdi() bool innerAboveThreshold(uchar aboveThreshold, ushort outerIdx)` and 
`static GPUdi() bool innerAboveThresholdInv(uchar aboveThreshold, ushort outerIdx)`

Could you for the time being change this such, that you don't use a function pointer? You can pass in another int template parameter, and do an if on that, which will be equally fast. I'm filing a bug report to clang for this function pointer support, but I do not know if they will be willing to fix it, since they might say it violates the OpenCL spec. Let's see.

**Comment by davidrohr** [comment]
I have also seen that you have a debug flag in some of the GPU code, i.e.
```
  if (debug) {
    printf("Looking around %d, %d\n", gpad, time);
  }
```
in NoiseSuppression.cxx.
Could you use a preprocessor macro instead, such that the GPU doesn't see the printf in the first place, irrespective of whether the compiler optimizes this away?

**Comment by fweig** [comment]
> @fweig : There are more problems in OpenCL with the function pointers. You have 2 of them:
> `static GPUdi() bool innerAboveThreshold(uchar aboveThreshold, ushort outerIdx)` and
> `static GPUdi() bool innerAboveThresholdInv(uchar aboveThreshold, ushort outerIdx)`
> 
> Could you for the time being change this such, that you don't use a function pointer? You can pass in another int template parameter, and do an if on that, which will be equally fast. I'm filing a bug report to clang for this function pointer support, but I do not know if they will be willing to fix it, since they might say it violates the OpenCL spec. Let's see.

The Opencl C++ specs state:
> function  pointers ... [are not supported by OpenCL C++]  unless  they  are  a  compile-time  constant expression

So declaring the function pointers as `constexpr` could also solve this?

**Comment by davidrohr** [comment]
@fweig : Give me a second to verify this. Please note that clang does not implement OpenCL C++, but it implements C++ for OpenCL, which is different for insane reasons...

**Comment by davidrohr** [comment]
I just tried it with the current master of clang 11, and it doesn't work. So for now, we have to work around it. Let's see how the clang people react. If they fix it, we can reverse the workaround.

**Comment by fweig** [comment]
I don't think it matters too much either way. It seems unlikely that we'll need more than those two functions anyway, unless the algorithm changes dramatically.

I pushed some changes. Hopefully this fixes all issues.

**Comment by davidrohr** [comment]
@fweig : Just out of curiosity. From the OP, do I understand correctly that you fixed the problem with the double peaks? I just ran over some data with and without this PR, and the number of clusters is exactly the same.

**Comment by fweig** [comment]
Yes, that should be fixed now. On my data it resulted in a few less clusters and my script doesn't find anymore duplicate clusters in the dumped data. This was an issue with floating point rounding. Maybe the problem never occured on your hardware in the first place?

**Comment by davidrohr** [comment]
@fweig : fyi: bug report and discussion on function pointers in clang c++ for opencl: https://bugs.llvm.org/show_bug.cgi?id=44788

**Comment by davidrohr** [comment]
Seems all fine now. Could you partially squash the fixup commits, then we can merge it.

**Comment by fweig** [comment]
Double checked performance: On my testset this PR didn't affect runtime.

If i compile the standalone tool with current `dev`, CUDA aborts because it runs out of memory:
```
CUDA Error: 2 / out of memory (/home/weiglhofer/alice/O2/GPU/GPUTracking/Base/cuda/GPUReconstructionCUDA.cu:314) 
CUDA Memory Allocation Error                                                                                     
GPU Tracker initialization failed                                                                                
Error initializing GPUReconstruction! 
```
Instead i rolled back on my brach to before my changes.

**Comment by davidrohr** [comment]
following this up offline.
In general, you can use the --memSize [nBytes] option to define the amount of GPU memory used.
You can try some values ~4GB -- 8 GB.
This will disable auto-detection, and should give you a consistently working setup, in case auto-detection fails.

**Pull request merged**